feat(17/WAKU2-RLN-RELAY): multiple acceptable roots (#539)
vacp2p/rfc: content/docs/rfcs/17/README.md

  - Oskar Thorén <oskar@status.im>
---

> The `17/WAKU2-RLN-RELAY` protocol is an extension of `11/WAKU-RELAY` which additionally provides spam protection using [Rate Limiting Nullifiers (RLN)](/spec/32). 

Reviewer: ```suggestion
The `17/WAKU2-RLN-RELAY` protocol is an extension of `11/WAKU2-RELAY` which additionally provides spam protection using [Rate Limiting Nullifiers (RLN)](/spec/32). 
```

This is to prevent a newly registered peer from spamming the system by messaging for all the past epochs. 
`max_epoch_gap` is a system parameter for which we provide some recommendations in section [Recommended System Parameters](#recommended-system-parameters).

**Merkle Root Validation**

> The routing peers MUST check whether the provided merkle root in the RateLimitProof is valid.

Reviewer 2: ```suggestion
The routing peers MUST check whether the provided Merkle root in the `RateLimitProof` is valid.
```

> It can do so by maintaining a local set of valid merkle roots, which consist of `AcceptableRootWindowSize` past roots.

Reviewer 2: ```suggestion
It can do so by maintaining a local set of valid Merkle roots, which consist of `AcceptableRootWindowSize` past roots.
```

> This allows peers which are not well connected to the network to be able to send messages, accounting for propagation delay.

Reviewer 3: Perhaps useful to be as precise as possible here on why "not well connected" could imply using an old merkle root.

> This allows peers which are not well connected to the network to be able to send messages, accounting for propagation delay.

Reviewer 2: It would be better to say to `account for network delay`, the specific reason for the window of acceptable root idea is that:
Due to the network delay, Nodes may not get the events emitted by the contract at the same time, hence won't update their trees at the same time. As such, there can be discrepancies in the Merkle tree root of the nodes in the network. 
The other reason is that until a message is published and then propagated in the entire network, the state of the group may change, hence the tree root may change. As such, at the time that the message is verified by other nodes, their group state might have advanced. Due to this, we need to consider a margin. This also occurs due to network delay.

Author: Agreed! Addressed in [ec3f8fa](https://github.com/vacp2p/rfc/pull/539/commits/ec3f8faea0e7b732ce73571834398606f8126e31)

> `AcceptableRootWindowSize` is a system parameter for which we provide some recommendations in section [Recommended System Parameters](#recommended-system-parameters)

Reviewer 2: The namings are not consistent `acceptable_root_window_size` and `AcceptableRootWindowSize`, please use one of them.

Author: This has been resolved

> This allows peers which are not well connected to the network to be able to send messages, accounting for propagation delay. This propagation delay is related to the nature of asynchronous network conditions, which means that peers see membership changes asynchronously, and therefore may have differing local merkle trees.

Reviewer 4: sembr (https://sembr.org/)

Author: Addressed in [3ba4c3c](https://github.com/vacp2p/rfc/pull/539/commits/3ba4c3c75182221f079dab3ec3f935a3464b6803)

> `acceptable_root_window_size` is a system parameter for which we provide some recommendations in section [Recommended System Parameters](#recommended-system-parameters)

Reviewer 4: I'd be more declarative here, e.g. ~See recommendations on how to choose this

Author: Addressed in [3ba4c3c](https://github.com/vacp2p/rfc/pull/539/commits/3ba4c3c75182221f079dab3ec3f935a3464b6803)

> | `staked_fund` | the amount of ether to be staked by peers at the registration |

Reviewer 5: Beyondt this PR but aren't we missing types here? Is this Ether? gwei, wei?

Author: Should be ether - cc: @staheri14 ?

Reviewer 2: The unit is wei (ether / 1e18)

| `reward_portion` | the percentage of `staked_fund` to be rewarded to the slashers |
| `max_epoch_gap` | the maximum allowed gap between the `epoch` of a routing peer and the incoming message |

> | `acceptable_root_window_size` | The maximum number of past Merkle roots to store |

Reviewer 5: Should we also include the type? i.e. `uint64`?

I can see in the table above we have it, i.e. `proof` is an array of 256 bytes.

Author: Hmm. since proof is sent over the wire we *do* need to define the type strictly. However, for something like `acceptable_root_window_size`, I do not think we need to define the type, and leave it up to the implementer.

## Epoch Length
A sensible value for the `period` depends on the application for which the spam protection is going to be used.
For example, while the `period` of `1` second i.e., messaging rate of `1` per second, might be acceptable for a chat application, might be too low for communication among Ethereum network validators.
One should look at the desired throughput of the application to decide on a proper `period` value.

> In the proof of concept implementation of `17/WAKU2-RLN-RELAY` protocol which is available in [nim-waku](https://github.com/status-im/nim-waku), the `period` is set to `1` second.

Reviewer 5: Looks like there is a redirect nim-waku->nwaku, but perhaps [nwaku](https://github.com/status-im/nwaku) to be consistent with the naming?

Author: I can create a follow up PR which renames all instances of nim-waku to nwaku. Thanks!

`Network_Delay` and `Clock_Asynchrony` MUST have the same resolution as  `period` .
By this formulation,  `max_epoch_gap`  indeed measures the maximum number of `epoch`s that can elapse since a message gets routed from its origin to all the other peers in the network.

`acceptable_root_window_size` depends upon the underlying chain's average blocktime, and `Network_Delay`

`acceptable_root_window_size` can be set as $acceptable_root_window_size=block_time/Network_Delay$

> By this formulation, `acceptable_root_window_size` will provide an approximation of how many roots can be acceptable by a routing peer.

Reviewer 4: I'd be slightly more declarative here

Author: Addressed in [3ba4c3c](https://github.com/vacp2p/rfc/pull/539/commits/3ba4c3c75182221f079dab3ec3f935a3464b6803)

> However, the peer can be generous with this window size if they desire enhanced slashing opportunities.

Reviewer 3: If different peers within the same spam-protected pubsub/content topic choose different window sizes, wouldn't this affect message routing quite significantly? Should this be a property agreed upon for the entire spam-protected domain?

Author: Indeed, this would affect message routing. However, due to the slashing incentive, peers would be attempting to maximize the number of roots they store locally (as long as the roots are within the maximum epoch time, otherwise the message is invalid nonetheless). Eventually, we will see all nodes converge upon a single value of `acceptable_root_window_size` to maximize opportunities to slash.

Additionally, There could be a scenario in which a malicious actor would set the `acceptable_root_window_size` to 0, but that would not have an impact on a well connected network, unless a majority attack is carried out, which is out of scope for this RFC.

Reviewer 5: Feel free to discuss this in a different channel, was wondering if we should have a separate spec for the `slasher`.

As I understand it, the merkle root is required for:
* 1) routing. a message containing an invalid one should be discarded
* 2) slashing. since it can be used as a proof to slash?

So if we define another protocol `WAKU2-RLN-SLASHER`, when setting the `acceptable_root_window_size` here in the `rln-relay`, perhaps we should only care about 1)

Then in the `slasher` protocol, we define all the logic to proof submiting to slash and get the reward, focusing on 2). This will allow anyone to run a lightweight slasher node.

Actually in ethereum consensus layer, the slasher is a different piece of software.

Slightly [related](https://discord.com/channels/864066763682218004/943703276132663296/1027169313867694110)

`Network_Delay` and `Clock_Asynchrony` MUST have the same resolution as  `period` .
By this formulation,  `max_epoch_gap`  indeed measures the maximum number of `epoch`s that can elapse since a message gets routed from its origin to all the other peers in the network.

`acceptable_root_window_size` depends upon the underlying chain's average blocktime, and `Network_Delay`

> `acceptable_root_window_size` SHOULD be set as $acceptable_root_window_size=block_time/Network_Delay$

Reviewer 2: Can you please elaborate on the reasoning behind this formula?  To me, it should be the other way around i.e., Net delay/ block time
if a messages takes X ms to reach the entire network, then we should measure that within that X span, how many blocks may get mined (due to which the tree state may get updated), then that number of mined blocks during X would be an approximation of the window size (it is a lower bound but not upper bound).
We also need to account for the block propagation delay in the Ethereum network as well, i.e.,  the time that it takes until all the Ethereum clients become aware of the new block.

Author: Completely right! I didn't think of it that way. Will address soon.

Author: Addressed in [dc3d476](https://github.com/vacp2p/rfc/pull/539/commits/dc3d476848c3ca73c3fa9c45fdca020b82d84a22)

Reviewer 2: Thanks for addressing it,  can you please explain why `block_propagation_delay` is in the denominator,  and please also add your reasoning about this to the RFC? it will make it easier for the readers to relate to the formulation as well.

Author: Initially I had thought that we were measuring the acceptable window of roots by number of blocks generated during the network delay. However, the network delay is very small compared to the block time of any blockchain,  and therefore the number of acceptable window of roots should be calculated based on the formula proposed by you.

Author: Addressed in b76f6be

Reviewer 2: For ease of explanation, let's assume `acceptable_root_window_size= block_propagation_delay` is already part of the `block_time`, simplifying the formula to `Network_delay/block_time`.

`Network_Delay` and `Clock_Asynchrony` MUST have the same resolution as  `period` .
By this formulation,  `max_epoch_gap`  indeed measures the maximum number of `epoch`s that can elapse since a message gets routed from its origin to all the other peers in the network.

> `acceptable_root_window_size` depends upon the underlying chain's average blocktime, and `Network_Delay`

Reviewer 2: ```suggestion
`acceptable_root_window_size` depends upon the underlying chain's average blocktime, and `block_time`
```

Author: blocktime has already been mentioned. Did you mean -
```suggestion
`acceptable_root_window_size` depends upon the underlying chain's average blocktime, `block_time`
```

Reviewer 2: Right, exactly what you said.

> `acceptable_root_window_size` SHOULD be set as $acceptable_root_window_size=(Network_Delay+block_propagation_delay)/block_time$

Reviewer 2: ```suggestion
The lower bound for the `acceptable_root_window_size` SHOULD be set as $acceptable_root_window_size=(Network_Delay)/block_time$
```

> `block_propagation_delay` is the time taken for the peers to receive the proposed block.

Reviewer 2: ```suggestion

```

> `Network_Delay` and `block_propagation_delay` MUST have the same resolution as `block_time`.

Reviewer 2: ```suggestion
`Network_Delay` MUST have the same resolution as `block_time`.
```

> By this formulation, `acceptable_root_window_size` will provide an approximation of how many roots can be acceptable by a routing peer.

Reviewer 2: ```suggestion
By this formulation, `acceptable_root_window_size` will provide a lower bound of how many roots can be acceptable by a routing peer.
```

> The `acceptable_root_window_size` should indicate how many blocks may have been mined during the time it takes for a peer to receive a packet. 

Reviewer 2: ```suggestion
The `acceptable_root_window_size` should indicate how many blocks may have been mined during the time it takes for a peer to receive a message. 
```

This formula represents a lower bound of the number of acceptable roots. 

> However, the peer can be generous with this window size if they desire enhanced slashing opportunities.

Reviewer 2: Let's remove this part and add a more precise suggestion in future PRs. 
```suggestion

```
